src/pl-gmp.c

```diff
@@ -1848,96 +1848,63 @@ mpq_to_double(mpq_t q)
- * the compiler doesn't  know about or the  compiler cannot gurarantee
- * it is used elsewhere may cause  inlining.  Therefore we use a union
- * and compare the integer component.
+ * Prolog implementation:
+ 
+rat_rationalize(Flt, Rat) :-
+   R is rational(Flt), rational(R, N, D),
+   rat_iter((N,D), (1,0), (0,1), Flt, Rat).
+
+rat_iter((V,W), (M,N), (P,Q), Flt, Rat) :-
+	divmod(V, W, D, U),
+	A is D*M+P,  %  A = Pnxt
+	B is D*N+Q,  %  B = Qnxt
+	Try is A rdiv B,
+	( (float(Try) =:= Flt ; U == 0)  % terminating conditions
+	  -> Rat = Try
+	  ;  rat_iter((W,U), (A,B), (M,N), Flt, Rat)
+	).
+
  */
 
 void
-mpq_set_double(mpq_t q, double f)	/* float -> nice rational */
-{ fpattern target = {.d = (f < 0.0) ? -f : f};
-  double x = target.d;
-  mpq_t b, c;
-  MP_INT *pna = mpq_numref(q);		/* use output q for a directly */
-  MP_INT *pda = mpq_denref(q);
-  MP_INT *pnb = mpq_numref(b);
-  MP_INT *pdb = mpq_denref(b);
-  MP_INT *pnc = mpq_numref(c);
-  MP_INT *pdc = mpq_denref(c);
-  mpz_t big_xi;
-  int half_exp;                       /* predict bitwise conversion size */
-  double fr = frexp(target.d, &half_exp);
-  int bitwise_denominator_size = DBL_MANT_DIG-(half_exp-1);
-
-  (void)fr;
-  if ( bitwise_denominator_size < 1 )
-    bitwise_denominator_size = 1;
-
-  mpz_set_ui(pna, 1L);                /* a = q = 1/0 */
-  mpz_set_ui(pda, 0L);
-  mpq_init(b);			      /* b = 0/1 */
-  mpq_init(c);			      /* auxiliary */
-  mpz_init(big_xi);                   /* auxiliary */
-
+mpq_set_double(mpq_t r, double f)	/* float -> nice rational */
+{ mpz_t m; mpz_init_set_si(m, 1);   // (m,n) = (1,0)
+  mpz_t n; mpz_init_set_si(n, 0);
+  mpz_t p; mpz_init_set_si(p, 0);   // (p,q) = (0,1)
+  mpz_t q; mpz_init_set_si(q, 1);
+
+  mpq_set_d(r, f);
+  mpz_t v; mpz_init(v); mpq_get_num(v,r);  // (v,w) == (r.num, r.den)
+  mpz_t w; mpz_init(w); mpq_get_den(w,r);
+
+  mpz_t d; mpz_init(d);
+  mpz_t u; mpz_init(u);
+  
   for(;;)
-  { double xf, xi;
-
-    /* infinite x indicates failure to converge */
-    if ( !isfinite(x) )
-      goto _bitwise_conversion_;
-
-    xf = modf(x, &xi);
-
-    /* compute a = a*xi + b for both numerator and denominator */
-    mpq_swap(q, b);
-    if ( x < (double)LONG_MAX )
-    { unsigned long int_xi = (unsigned long) xi;
-      mpz_mul_ui(pnc, pnb, int_xi);
-      mpz_mul_ui(pdc, pdb, int_xi);
-    } else
-    { mpz_set_d(big_xi, xi);
-      mpz_mul(pnc, pnb, big_xi);
-      mpz_mul(pdc, pdb, big_xi);
-    }
-    mpz_add(pna, pna, pnc);
-    mpz_add(pda, pda, pdc);
-
-    /* if it gets too long, fall back to bitwise conversion */
-    if (mpz_sizeinbase(pda, 2) > bitwise_denominator_size)
-      goto _bitwise_conversion_;
-
-    /* See (*) above */
-    fpattern iter = {.d = mpz_fdiv(pna, pda)};
-    if ( target.i == iter.i )
+  { mpz_fdiv_qr(d,u,v,w);
+    mpz_addmul(p,d,m);
+    mpz_addmul(q,d,n);
+    if ((mpz_fdiv(p,q) == f) || (mpz_sgn(u) == 0)) {  // terminating conditions
```

Comment:
I only have a very superficial understanding of what you're talking about. I assume that this is an issue because `mpz_fdiv` is `static` and part of the same compilation unit, so the compiler knows the return value is in an extended float register?

Would it make any difference if either a) `mpz_fdiv` wasn't `static`, or b) the result got put into a local `double` and that value was returned? 

That said,  would it be sufficient to put  the result of `mpz_fdiv`  `fpattern` and then test against `f` (the input argument `double`):
```
   fpattern iter;
 .
 .
    iter.d = mpz_fdiv(p,q);
    if ((iter.d == f) || (mpz_sgn(u) == 0)) {  // terminating conditions  
 .
 .  
```
or will the compiler see through this "ruse" as well?

Reply:
Hard to say.   Compilers get smarter.   This was never a problem, but it became one with GCC 12 on i386 architectures (could reproduce on a VM on an ADM64 machine).   Assuming the compiler doesn't see through is typically a bad idea.   There is more and more stress on specializing functions, link-time optimization, etc.    By comparing the int64 member it __has__ to place the double into the union and __has__ to compare 64 bits.    I don't think it has much runtime overhead as all should remain in registers.   It merely forces the compiler to put the value in a normal double register rather than the extended precision float register. The alternative is to place the float in a `volatile` location that is publically visible (so it can't be optimized), but that involves transfer to main memory which probably is slower.

If you see something better, be my guest.   Otherwise please restore the hack :cry:

Reply:
OK, I think I've updated the PR to include this, but not sure how to test that it does the job. Please review to make sure I got it right. 

Seems to me that this is a compiler optimization bug; the fact that it has persisted is just another reason why C (and family) are at the bottom of my favourite languages list.